encoders/firebase-decoders-json/src/test/java/com/google/firebase/decoders/json/JsonDataDecoderExtraPropertyTest.java

```diff
@@ -0,0 +1,126 @@
+import com.google.firebase.decoders.FieldModifier;
+import com.google.firebase.decoders.FieldRef;
+import com.google.firebase.decoders.ObjectDecoder;
+import com.google.firebase.decoders.ObjectDecoderContext;
+import com.google.firebase.decoders.Safe;
+import com.google.firebase.decoders.TypeCreator;
+import com.google.firebase.decoders.TypeToken;
+import com.google.firebase.encoders.FieldDescriptor;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.annotation.Annotation;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.util.HashMap;
+import java.util.Map;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+
+@RunWith(RobolectricTestRunner.class)
+public class JsonDataDecoderExtraPropertyTest {
+
+  @Retention(RetentionPolicy.RUNTIME)
+  @Target({ElementType.METHOD, ElementType.FIELD})
+  private @interface Default {
+    String value();
+
+    Default INSTANCE = new DefaultImpl();
+
+    class DefaultImpl implements Default {
+      @Override
+      public Class<? extends Annotation> annotationType() {
+        return Default.class;
+      }
+
+      @Override
+      public String value() {
+        return "default";
+      }
+
+      @Override
+      public boolean equals(Object obj) {
+        return super.equals(obj);
+      }
+
+      @Override
+      public int hashCode() {
+        return super.hashCode();
+      }
+    }
+  }
+
+  private FieldModifier<Default> processor =
+      new FieldModifier<Default>() {
+        @Nullable
+        @Override
+        public <T> T apply(
+            @NonNull Default annotation, @Nullable T fieldDecodedResult, @NonNull Class<T> type) {
+          if (fieldDecodedResult == null) {
+            if (type.equals(String.class)) return (T) annotation.value();
+          }
+          return fieldDecodedResult;
+        }
+      };
+
+  static class Foo {
+    @Default("default")
+    String str;
+
+    Foo(String str) {
+      this.str = str;
+    }
+  }
+
+  static class FooObjectDecoder implements ObjectDecoder<Foo> {
+    @NonNull
+    @Override
+    public TypeCreator<Foo> decode(@NonNull ObjectDecoderContext<Foo> ctx) {
+      FieldDescriptor strFieldDescriptor =
+          FieldDescriptor.builder("str").withProperty(Default.INSTANCE).build();
+      FieldRef.Boxed<String> strField = ctx.decode(strFieldDescriptor, TypeToken.of(String.class));
+      return (creationCtx -> new Foo(creationCtx.get(strField)));
+    }
+  }
+
+  @Test
+  public void customizedAnnotation_shouldProcessCorrectly() throws IOException {
+    Map<Class<?>, ObjectDecoder<?>> objectDecoders = new HashMap<>();
+    Map<Class<?>, FieldModifier<?>> processors = new HashMap<>();
+
+    objectDecoders.put(Foo.class, new FooObjectDecoder());
+    processors.put(Default.class, processor);
```

Comment:
I think you can just replace the processor with an inline lambda expression here, wdyt?

Reply:
just tried to place them in an inline lambda expression, but seems like the compiler won’t be able to determine those generic types where we have to cast them multiple times, will use an anonymous class instead.